pkg/execution/pauses/iterator.go

```diff
@@ -0,0 +1,244 @@
+	// have older pauses loaded when the buffer iteration has finished.
+	//
+	// Iteration does NOT need to happen in-order.
+	d.fetchNextBlocks()
+
+	// Check on the buffer iterator and advance on this.
+	if d.bufferIter.Next(ctx) {
+		d.usingBuffer = true
+		return true
+	}
+
+	d.usingBuffer = false
+
+	// NOTE: We must release the lock as soon as possible such that the fetchBlock
+	// background thread can grab the lock to adjust pauses.
+	d.l.Lock()
+	quit := len(d.pauses) == 0 && len(d.inflightBlocks) == 0
+	d.l.Unlock()
+
+	if quit {
+		// We are done!  There are no pauses downloaded or inflight.
+		return false
+	}
+
+	// Complex case:  we may be waiting for in flight blocks to download, but only
+	// if we've handled all already downloaded pauses:
+	//
+	// eg:  0 blocks downloaded / 10 being downloaded.  spin until len(d.pauses) > 0.
+	d.l.Lock()
+	spin := len(d.inflightBlocks) > 0 && len(d.pauses) == 0
+	d.l.Unlock()
+
+	for spin {
+		// Wait 100ms for the block to download and try again.
+		time.Sleep(100 * time.Millisecond)
+
+		d.l.Lock()
+		if d.err != nil {
+			d.l.Unlock()
+			// Skip as we've errored.
+			return false
+		}
+
+		spin = len(d.inflightBlocks) > 0 && len(d.pauses) == 0
+		d.l.Unlock()
+	}
+
+	// Now we're good.
+	d.l.Lock()
+	defer d.l.Unlock()
+	if len(d.pauses) > 0 {
+		// Simple case:  pauses are downloaded, so yeah we have a next value.
+		return true
+	}
+
+	// If we're here and we've already processed all in-flight blocks, attempt
+	// to redownload blocks.
+	return d.Next(ctx)
+}
+
+// Error returns the error returned during iteration, if any.  Use this to check
+// for errors during iteration when Next() returns false.
+func (d *dualIter) Error() error {
+	return d.err
+}
+
+// Val returns the current Pause from the iterator.
+func (d *dualIter) Val(ctx context.Context) *state.Pause {
+	atomic.AddInt64(&d.index, 1)
+
+	if d.usingBuffer {
+		return d.bufferIter.Val(ctx)
+	}
+
+	d.l.Lock()
+	defer d.l.Unlock()
+
+	if len(d.pauses) == 0 {
+		return nil
+	}
+
+	pause := d.pauses[0]
+	d.pauses = d.pauses[1:]
+
+	return pause
+}
+
+// Index shows how far the iterator has progressed
+func (d *dualIter) Index() int64 {
+	return d.index
+}
+
+// fetchNextBlocks fetches blocks, returning whether we're fetching anything
+// in the background.
+func (d *dualIter) fetchNextBlocks() bool {
+	d.l.Lock()
+	defer d.l.Unlock()
+
+	maxFetch := DefaultConcurrentBlockFetches - len(d.inflightBlocks)
+	if maxFetch > len(d.unfetchedBlocks) {
+		maxFetch = len(d.unfetchedBlocks)
+	}
+
+	if maxFetch == 0 {
+		return false
+	}
+
+	blockIDs := d.unfetchedBlocks[0:maxFetch]
+
+	// Move the blocks to in-flight.
+	d.inflightBlocks = append(d.inflightBlocks, blockIDs...)
+
+	// And remove from unfetched
+	d.unfetchedBlocks = d.unfetchedBlocks[maxFetch:]
+
+	for _, bid := range blockIDs {
+		// Fetch each block, without capturing the ID.
+		id := bid
+		go d.fetchBlock(context.Background(), id)
```

Comment:
should this be done in a pool?

Reply:
We only fetch up to 20 at a time so I don't think it's worth it.